### ISSUE TITLE
ValueError: Specified GPU device id (0) is invalid

### ISSUE THREAD

**Issue by youssefhb** [Microsoft/CNTK#1835]
Hi,
i have built CNTK GPU successfully, and i can run the FeedForwardNet.py example.
However when i try to run the same example using GPU (try_set_default_device(cntk.device.gpu(0)))
I get the following error :   

Traceback (most recent call last):
  File "FeedForwardNet.py", line 82, in <module>
    try_set_default_device(cntk.device.gpu(0))
  File "/CNTK/bindings/python/cntk/internal/swig_helper.py", line 69, in wrapper
    result = f(*args, **kwds)
  File "/CNTK/bindings/python/cntk/device.py", line 94, in gpu
    return cntk_py.DeviceDescriptor.gpu_device(device_id)
ValueError: Specified GPU device id (0) is invalid.

Build info:

		Built time: May  1 2017 22:44:50
		Last modified date: Mon May  1 12:04:32 2017
		Build type: release
		Build target: GPU
		With 1bit-SGD: no
		With ASGD: no
		Math lib: mkl
		CUDA_PATH: /usr/local/cuda-8.0/
		CUB_PATH: /usr/local/cub-1.4.1
		CUDNN_PATH: /usr/local/cudnn-5.1
		Build Branch: master
		Build SHA1: 4b9f8739c72068d70279f91b4b59923b2ae1fc3a (modified)
	
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 375.39                 Driver Version: 375.39                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 675M    Off  | 0000:01:00.0     N/A |                  N/A |
|  0%   59C    P0    N/A /  N/A |    357MiB /  1984MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0                  Not Supported                                         |
+-----------------------------------------------------------------------------+

Anyone has faced this issue ?
Thanks.

**Comment by Sadler2** [comment]
I have exactly the same error. I just installed https://cntk.ai/PythonWheel/GPU/cntk-2.0rc2-cp35-cp35m-win_amd64.whl on my Windows 7 x64, Anaconda 3 4.1.1 (64 bit)

SMI:
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 368.81                 Driver Version: 368.81                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name            TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 660    WDDM  | 0000:01:00.0     N/A |                  N/A |
| 13%   40C    P8    N/A /  N/A |    192MiB /  2048MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
```

Build info:
```
 Built time: Apr 21 2017 01:32:20
 Last modified date: Fri Apr 21 00:23:15 2017
 Build type: Release
 Build target: GPU
 With 1bit-SGD: no
 With ASGD: yes
 Math lib: mkl
 CUDA_PATH: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v8.0
 CUB_PATH: c:\local\cub-1.4.1
 CUDNN_PATH: C:\local\cudnn-8.0-windows10-x64-v5.1
 Build Branch: HEAD
 Build SHA1: f190457aecbfe07c3fe27c89f36108bc2488727b
 Built by svcphil on liana-08-w
 Build Path: C:\jenkins\workspace\CNTK-Build-Windows\Source\CNTKv2LibraryDll\
 MPI distribution: Microsoft MPI
 MPI version: 7.0.12437.6
```
`cntk.device.all_devices()` contains CPU only. I have also tried to replace the default .dll's from package with my local CUDA 8 and cuDNN 5.1 ones, it didn't help.

**Comment by Ark-kun** [comment]
I have the same problem.
It's random for me.
Usually, it works, but when I have MS Edge open, it usually does not work.

**Comment by miguelgfierro** [comment]
It looks that you have installed the CPU version instead of the GPU version. I would recommend to reinstall the last GPU version and check that CNTK is using GPUs:
```python
import cntk
cntk.logging.set_trace_level(2)
cntk.all_devices() 
```

**Comment by youssefhb** [comment]
From build info, you can see that build target is GPU.
GPU is working fine with other libraries. The issue in CNTK side.

**Comment by miguelgfierro** [comment]
If you print `cntk.all_devices()` do you get CPU or GPU?

**Comment by youssefhb** [comment]
it's just empty.

**Comment by miguelgfierro** [comment]
That's odd. At least you should get CPU. I would try to install the last release RC3. I just installed this version today in a windows GPU VM and it worked.

**Comment by youssefhb** [comment]
I'll try the latest release. Hope it will resolve my issue.
Thank you.

**Comment by KeDengMS** [comment]
@youssefhb is this issue resolved?

**Comment by cha-zhang** [comment]
I'm closing this issue due to lack of response. Assume this is resolved for the latest 2.0 release. Feel free to reopen this issue if needed. Thanks!

**Comment by bpatil16** [comment]
I have install CNTK-2-0-Windows-64bit-GPU on windows 10, when I run  cntk.all_devices() I get CPU, how to activate GPU. any answers

**Comment by KeDengMS** [comment]
Usually this is a CUDA issue. Please install cuda samples and run device query sample to make sure CUDA works.

**Comment by bpatil16** [comment]
looks like even if I install the cuda I dont have  GPU that CUDA-capable. I followed this

2.1. Verify You Have a CUDA-Capable GPU
To verify that your GPU is CUDA-capable, open the Control Panel ( Start > Control
Panel ) and double click on System. In the System Properties window that opens, click
the Hardware tab, then Device Manager. Expand the Display adapters entry. There you
will find the vendor name and model of your graphics card. If it is an NVIDIA card that
is listed in http://developer.nvidia.com/cuda-gpus, your GPU is CUDA-capable.
The Release Notes for the CUDA Toolkit also contain a list of supported products

**Comment by bpatil16** [comment]
I also dont have a NIVIDIA graphic card on my machine.

**Comment by giuliohome** [comment]
I am still having this issue.
My GPU card is GeForce GT 540M (capability 2.1)
Can you reopen the issue and give support to solve it please?

**Comment by giuliohome** [comment]
I've tried CUDA 8.0. The Visual Studio Integration (I have vs2015) **fails**.
Aside from that I have tried many version of CNTK starting from 2.0 and I couldn't make it work... I will test with CNTK 2.3 again and let you know in case I have any postive result

**Comment by giuliohome** [comment]
@KeDengMS 
This is the CNTK installed
```
(base) C:\Users\MyHome>pip install https://cntk.ai/PythonWheel/GPU/cntk-2.3-cp36
-cp36m-win_amd64.whl
Collecting cntk==2.3 from https://cntk.ai/PythonWheel/GPU/cntk-2.3-cp36-cp36m-wi
n_amd64.whl
  Downloading https://cntk.ai/PythonWheel/GPU/cntk-2.3-cp36-cp36m-win_amd64.whl
(311.0MB)
    100% |¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦| 311.0MB 18kB/s
Requirement already satisfied: scipy>=0.17 in c:\dev\python\jupiter\conda\lib\si
te-packages (from cntk==2.3) (0.19.1)
Requirement already satisfied: numpy>=1.11 in c:\dev\python\jupiter\conda\lib\si
te-packages (from cntk==2.3) (1.13.3)
Installing collected packages: cntk
Successfully installed cntk-2.3
```
printing the version as expected
```
(base) C:\Users\MyHome> python -c "import cntk; print(cntk.__version__)"
2.3
```
I can't still have the GPU recognized as a valid device
```
(base) C:\Users\MyHome>python -c "import cntk; cntk.logging.set_trace_level(2);
cntk.all_devices()"
-------------------------------------------------------------------
GPU info:

                Device[0]: cores = 96; computeCapability = 2.1; type = "GeForce
GT 540M"; total memory = 2048 MB; free memory = 1990 MB
-------------------------------------------------------------------
```

**Comment by giuliohome** [comment]
Same issue from a trivial C# console app with Visual Studio 2015 after 
```
PM> Install-Package CNTK.GPU -Version 2.3.0


Attempting to gather dependency information for package 'CNTK.GPU.2.3.0' with respect to project 'ConsoleCNTK', targeting '.NETFramework,Version=v4.6.2'
Gathering dependency information took 1,23 sec
Attempting to resolve dependencies for package 'CNTK.GPU.2.3.0' with DependencyBehavior 'Lowest'
Resolving dependency information took 0 ms
Resolving actions to install package 'CNTK.GPU.2.3.0'
Resolved actions to install package 'CNTK.GPU.2.3.0'
Retrieving package 'CNTK.GPU 2.3.0' from 'nuget.org'.
  GET https://www.nuget.org/api/v2/package/CNTK.GPU/2.3.0
  OK https://www.nuget.org/api/v2/package/CNTK.GPU/2.3.0 1204ms
Installing CNTK.GPU 2.3.0.
Adding package 'CNTK.GPU.2.3.0' to folder 'C:\CNTK\mySample\ConsoleCNTK\packages'
Added package 'CNTK.GPU.2.3.0' to folder 'C:\CNTK\mySample\ConsoleCNTK\packages'
Added package 'CNTK.GPU.2.3.0' to 'packages.config'
Successfully installed 'CNTK.GPU 2.3.0' to ConsoleCNTK
Executing nuget actions took 5,58 min
Time Elapsed: 00:05:36.3834711
```

![VS exception in C#](https://snag.gy/IzJalQ.jpg)

Again, `DeviceDescriptor.AllDevices()` returns only one device `[0]` of  `.Type` equals to `CPU`

**Comment by giuliohome** [comment]
I'm afraid that CNTK 2.3 does *not* support GPU with _compute capability_ 2.1 for CUDA (and not even CNTK 2.0)
see this [line](https://github.com/Microsoft/CNTK/blob/release/2.3/Source/CNTK/CNTK.cpp#L135) from the source of the branch `release/2.3`
`CNTK: The GPU (%s) has compute capability %d.%d.  CNTK is only supported on GPUs with compute capability 3.0 or greater"`

**Comment by KeDengMS** [comment]
Please try a newer GPU then, capability 2.x is behind us now.

**Comment by KeDengMS** [comment]
@giuliohome I understood your frustration when the support of your hardware is dropped. If CNTK 2.3 does not work either, please try [older versions](https://docs.microsoft.com/en-us/cognitive-toolkit/setup-linux-python?tabs=cntkpy20). 
The fact is that NVidia has stopped supporting capability 2.x in CUDA 9.0, and now most toolkits have upgraded to CUDA 9.x as CNTK. If you feel strongly against NVidia's decision, please send your comments to them directly.

**Comment by nimning** [comment]
My cntk can not detect the GPU. It also returns (CPU, ). The following is the information of my CNTK and the CUDA? All the above cases do not seem to fall in my situation. Help?


`print(C.__version__)`
2.6

```
from numba import cuda
cuda.detect()
```
Found 4 CUDA devices
id 0           b'Tesla K40c'                              [SUPPORTED]
                      compute capability: 3.5
                           pci device id: 0
                              pci bus id: 4
id 1           b'Tesla K40c'                              [SUPPORTED]
                      compute capability: 3.5
                           pci device id: 0
                              pci bus id: 2
id 2           b'Tesla K40c'                              [SUPPORTED]
                      compute capability: 3.5
                           pci device id: 0
                              pci bus id: 131
id 3           b'Tesla K40c'                              [SUPPORTED]
                      compute capability: 3.5
                           pci device id: 0
                              pci bus id: 132
Summary:
	4/4 devices are supported

```
C.logging.set_trace_level(2);
C.all_devices()
```
(CPU,)

**Comment by nimning** [comment]
> Did you notice any warning in import cntk? Maybe some cuda libs are not found.

C:\ProgramData\Anaconda3\lib\site-packages\cntk\cntk_py_init.py:32: UserWarning: Unsupported Windows version (2012serverr2). CNTK supports Windows 10 and above, only.
  warnings.warn('Unsupported Windows version (%s). CNTK supports Windows 10 and above, only.' % __my_distro_ver__)

**Comment by GuntaButya** [comment]
I had this issue. I resolved my problem by ensuring the package CNTK.CPUOnly was **not installed**

![cntk cpu only](https://user-images.githubusercontent.com/24576124/52826255-d8033180-310b-11e9-9cb7-6c69cbc87ab7.png)


This somehow killed the GPU component:

![cntk gpu with cpu only](https://user-images.githubusercontent.com/24576124/52826770-fb2ee080-310d-11e9-8a8e-e0e8f902dff4.png)


I uninstalled the whole lot, reinstalled just: **CNTK.GPU**

Run:

>          var device = CNTK.DeviceDescriptor.GPUDevice(0);
            Console.WriteLine($"Using:  {device.Type} Device info: {device.AsString()}");
            Console.ReadLine();

and the problem was resolved!

Its worth noting, the issue I saw was on the CNTK Examples downloaded from the CNTK GitHub site. Restoring the packages is the only thing I can think of that may caused the issue.
